### PR TITLE
Adds beforeFind to RelatedModelsListener, for View

### PR DESCRIPTION
**Pull request by eaxexe**
Change makes it possible to get related models for Crud.View, just like we have with Crud.Index, making it more streamlined.

refs #402

### CODE REVIEW

**Comment by eaxexe** [comment]
I am aware that the checks fail - However, if you pull crud and run phpunit currently, you will get the failed test, that also fails here.

The test I have added works, when I run it :)

**Comment by ADmad** [comment]
This change will overlaps with [ViewListener::beforeFind()](https://github.com/FriendsOfCake/crud-view/blob/master/src/Listener/ViewListener.php#L29) which does similar thing but provides more control over the models to contain.

If this `RelatedModelsListener` runs before `ViewListener` one will lose the customizability the latter provides (since it bails out if containments already already set) and cause problems with existing app.

**Comment by eaxexe** [comment]
It seems to me, that the functionality should be in crud rather than crud-view, since many projects, like mine, may want to use crud without crud-view.

Beyond that, it is a streamlining of relatedModels<http://crud.readthedocs.io/en/latest/listeners/related-models.html> in crud, that Index and View behave the same, causing less confusion.

If a developer uses crud-view, and prefers ViewListener::beforeFind() <https://github.com/FriendsOfCake/crud-view/blob/master/src/Listener/ViewListener.php#L29> from that, then it is a simple matter of **not** running:
`$this->Crud->listener('relatedModels')->relatedModels([‘Articles’], 'view');`
and thus avoiding the collision, or is that an incorrect understanding?

Since this is an addition and not a change, I don’t see how it can cause problems with existing apps, unless they have run:
$this->Crud->listener('relatedModels')->relatedModels(['Areas'], 'view');
while that part was defunct?

**Comment by ADmad** [comment]
> it is a streamlining of relatedModels in crud, that Index and View behave the same, causing less confusion.

I agree. I am not opposed to this change but we need to consider the possible implication on existing apps.

> I don’t see how it can cause problems with existing apps, unless they have run:
$this->Crud->listener('relatedModels')->relatedModels(['Areas'], 'view');

Well that's not the only way to attach listeners. One can also use `'listeners'` key in `CrudComponent` option and those listeners will be loaded for all actions. So if someone has `Crud.RelatedModels` before `CrudView.View` in the listeners (which is very likely), after this change the contains will be done by the former and the latter will bail out since it checks whether contains are already set.

So we need to think of a way to add this feature without potentially affecting CrudView users.

File: src/Listener/RelatedModelsListener.php [hunk 1]
```diff
@@ -51,6 +52,27 @@ public function beforePaginate(Event $event)
...
 
     /**
+     * Automatically parse and contain related table classes
+     *
+     * @param \Cake\Event\Event $event Before find event
+     * @return void
+     */
+    public function beforeFind(Event $event)
+    {
+        $contained = $event->subject()->query->contain();
```
Comment (ADmad) [inline src/Listener/RelatedModelsListener.php]: Instead of this code duplication you can just call `$this->beforePaginate()`

Comment (ADmad) [inline src/Listener/RelatedModelsListener.php]: Or make a put the common code in a protected function `_eventHandler()` and call that from both the callback methods.

Comment (eaxexe) [inline src/Listener/RelatedModelsListener.php]: Fantastic idea! Should make the code simpler and clearer, I have implemented it and am pushing it in a bit.

**Comment by ayman-alkom** [comment]
+1

**Comment by jippi** [comment]
Stalled for a year :)

**Pull request closed without merging**